Box documentation (#1728)
elastic/connectors-python: docs/reference/box.md

# Box Connector

The [Elastic Box connector](../connectors/sources/box.py) is built with the Elastic connectors Python framework and is available as a self-managed [connector client](https://www.elastic.co/guide/en/enterprise-search/current/build-connector.html).

## Availability and prerequisites

This connector is available as a **connector client** from the **Python connectors framework**. To use this connector, satisfy all [connector client requirements](https://www.elastic.co/guide/en/enterprise-search/master/build-connector.html).

This connector is in **beta** and is subject to change. The design and code is less mature than official GA features and is being provided as-is with no warranties. Beta features are not subject to the support SLA of official GA features.

## Usage

To use this connector as a **connector client**, use the **Box** tile from the connectors list OR **Customized connector** workflow.

For additional operations, see [Usage](https://www.elastic.co/guide/en/enterprise-search/master/connectors-usage.html).

## Box API Authorization

### Box Free Account

#### Create Box User Authentication (OAuth 2.0) Custom App

> You'll need to create an User Authentication (OAuth 2.0) app in the Box developer console by following these steps:

Reviewer: ```suggestion
You'll need to create an OAuth app in the Box developer console by following these steps:
```

1. Register a new app in the [Box dev console](https://app.box.com/developers/console) with custom App and select User authentication (OAuth 2.0).

> 2. Add your Elasticsearch endpoint URL to **Redirect URIs**.

Reviewer 2: Why Elasticsearch endpoint?

Reviewer 3: because it will redirect to user where he/she needs to configure the connector.

Author: Because it will redirect the user to where he or she needs to configure the connector.

Reviewer 2: I'm afraid the Elasticsearch endpoint can't be an HTML page. Elasticsearch exposes only API afaik. 

Also, how can the customer grab an `authorization code` if they are redirected to another page? Usually, a page should handle the `authorization code` and use it.

Author: Okay then I have removed this step and now user will redirect to default location

Reviewer 2: Can you please describe how customers can find `authorization code`? 

As far as I understand, the authorization code is a part of the redirect payload so the customers should be able to see it. We need to list those steps. 

I was able to grab the code only when I created a local web server and used localhost as a redirect page. Do you have other options?

Author: I have updated the doc to find `authorization code`

> 2. Add elasticsearch index URL in Redirect URIs

Reviewer: ```suggestion
2. Add your Elasticsearch endpoint URL to **Redirect URIs**.
```

3. Check "Write all files and folders stored in Box" in Application Scopes.

> 4. Once the app is created, users can have Client ID and Client secret on the configuration tab.

Reviewer: ```suggestion
4. Once the app is created, **Client ID** and **Client secret** values are available in the configuration tab. Keep these handy.
```

#### Generate a refresh Token

To generate a refresh token, follow these steps:

1. Go to the following URL, replacing `<CLIENT_ID>` with the **Client ID** value saved earlier.
    ```shell

>     "https://account.box.com/api/oauth2/authorize?response_type=code&client_id=<CLIENT_ID>"

Reviewer: ```suggestion
    https://account.box.com/api/oauth2/authorize?response_type=code&client_id=<CLIENT_ID>
```

Don't think need quotes here?

    ```
    
The HTTP response should contain an **authorization code** that you'll use to generate a refresh token.

> **Note:** Authorization code can only be used once to create a refresh token and itâ€™s valid for 30 seconds only.

Reviewer: ```suggestion
**Note:** Authorization codes to generate refresh tokens can only be used once and are only valid for 30 seconds.
```

> 2. In your terminal, run the following `cURL` command, replacing `<AUTHORIZATION_CODE>`, `<CLIENT_ID> and <CLIENT_SECRET>` with the values you saved earlier:

Reviewer: ```suggestion
2. In your terminal, run the following `curl` command, replacing `<AUTHORIZATION_CODE>`, `<CLIENT_ID> and <CLIENT_SECRET>` with the values you saved earlier:
```

    ```shell
    curl -i -X POST "https://api.box.com/oauth2/token" \
        -H "Content-Type: application/x-www-form-urlencoded" \
        -d "client_id=<CLIENT_ID>" \
        -d "client_secret=<CLIENT_SECRET>" \
        -d "code=<AUTHORIZATION_CODE>" \
        -d "grant_type=authorization_code"
    ```

> Store the refresh token from the response and use it in the connector configuration.

Reviewer: ```suggestion
Save the refresh token from the response. You'll need this for the connector configuration.
```

### Box Enterprise Account

#### Create Box Server Authentication (Client Credentials Grant) Custom App

1. Register a new app in the [Box dev console](https://app.box.com/developers/console) with custom App and select Server Authentication (Client Credentials Grant).

2. Check following permissions:
    - "Write all files and folders stored in Box" in Application Scopes
    - "Make API calls using the as-user header" in Advanced Features

3. Select `App + Enterprise Access` in App Access Level.

> 4. Then authorized your application by admin console.

Reviewer: ```suggestion
4. Authorize your application from the admin console.
```

> Store the Client Credentials and Enterprise ID and use it in the connector configuration.

Reviewer: ```suggestion
Save the **Client Credentials** and **Enterprise ID**. You'll need these to configure the connector.
```

## Configuration

### Configure Box connector

#### `Box Account`  (required)

Dropdown to determine Box Account type: `Box Free Account` or `Box Enterprise Account`. Default value is `Box Free Account`.

#### `Client ID`  (required)

The Client ID to authenticate with Box instance.

#### `Client Secret`  (required)

The Client Secret to authenticate with Box instance.

#### `Refresh Token` (required if Box Account is Box Free)

The Refresh Token to generate Access Token. 

> **NOTE:** If the process terminates, then user needs to generate a new refresh token and configure

Reviewer: ```suggestion
**NOTE:** If the process terminates, you'll need to generate a new refresh token.
```

    ```shell
    https://account.box.com/api/oauth2/authorize?response_type=code&client_id=<CLIENT_ID>
    ```

2. Grant access to your application.

> 3. Now you will redirect to the web page that you configured in **Redirect URIs**, and the HTTP response should contain an **authorization code** that you'll use to generate a refresh token.

Reviewer 2: ```suggestion
3. You will now be redirected to the web page that you configured in **Redirect URIs**, and the HTTP response should contain an **authorization code** that you'll use to generate a refresh token.
```
